### PR TITLE
Custom airgap

### PR DESCRIPTION
**Pull request by areed**
All supported versions of Kubernetes, Docker and addons are bundled into
individual tar archives in CI and uploaded to S3. When a new installer
is created, the server will create a Job to generate an airgap bundle.
The server passes the packages required for the installer to the Job
along with a signed URL for S3 where it can upload the bundle. The Job
is a simple bash script that downloads all the packages and reassembles
into a single archive.

The online install script uses the same packages used in the airgap
bundle but downloads them at runtime. The online server redirects all
dist/ (pacakge) and bundle/ requests to S3.

Future work:
- For new named installers the bundle may exist with a different ID. It
would be more efficient to re-use bundles when available.
- After POSTing a new installer it takes a minute for the airgap bundle
to be ready.
- There's no reconciler that detects failed airgap bundles and retries
or escalates.
- No garbage collection for ConfigMaps and Jobs.

### CODE REVIEW

File: bin/create-bundle-alpine.sh [hunk 1]
```diff
@@ -0,0 +1,24 @@
+#!/bin/sh
```
Comment (areed) [inline bin/create-bundle-alpine.sh]: This is the script that creates custom airgap bundles at runtime. The server adds it to a ConfigMap that the Job mounts. The server then passes the signed PUT url as the first arg and a list of URLs to individual packages to include.

File: bin/docker-save.sh [hunk 2]
```diff
@@ -0,0 +1,15 @@
+#!/bin/bash
```
Comment (areed) [inline bin/docker-save.sh]: This runs in CI to save docker images as tar archives so they can be included in individual packages.

File: bin/upload-dist.sh [hunk 3]
```diff
@@ -0,0 +1,13 @@
+#!/bin/bash
```
Comment (areed) [inline bin/upload-dist.sh]: This uploads all the individual packages to S3 during CI. Online installs download at runtime and airgap bundles download these when the Job runs.

File: bin/upload-latest-bundle.sh [hunk 4]
```diff
@@ -0,0 +1,39 @@
+#!/bin/bash
```
Comment (areed) [inline bin/upload-latest-bundle.sh]: This runs in CI to create the special airgap bundle named latest.

File: kustomize/base/namespace.yaml [hunk 5]
```diff
@@ -0,0 +1,4 @@
+apiVersion: v1
```
Comment (areed) [inline kustomize/base/namespace.yaml]: All the airgap Jobs run in this namespace so it's easier to keep track of what's been created.

File: packages/kubernetes/1.15.0/Manifest [hunk 6]
```diff
@@ -0,0 +1,4 @@
+kubernetes gcr.io/google-containers/hyperkube:v1.15.0
```
Comment (areed) [inline packages/kubernetes/1.15.0/Manifest]: List of images to include in the package. This is parsed in CI.

File: scripts/Manifest [hunk 7]
```diff
@@ -0,0 +1,8 @@
+KUBERNETES_VERSION=1.15.2
```
Comment (areed) [inline scripts/Manifest]: Latest manifest. This is turned into a template for the server to serve other install scripts.

File: web/src/controllers/Scripts.ts [hunk 8]
```diff
@@ -1,125 +1,86 @@
...
+
+      let installer = await this.installerStore.getInstaller(installerID);
+      if (!installer) {
         response.status(404);
         return notFoundResponse;
       }
-      installer = i;
-    } catch (error) {
-      return { error };
+      installer = installer.resolve();
```
Comment (areed) [inline web/src/controllers/Scripts.ts]: replace "latest" with actual versions

File: web/src/kubernetes/index.ts [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+    private readonly templates: Templates,
+  ) {
+    this.distOrigin = `https://${process.env["KURL_BUCKET"]}.s3.amazonaws.com`;
+    this.ns = process.env["KURL_NAMESPACE"] || "kurl";
+    this.disabled = !!process.env["AIRGAP_BUNDLING_DISABLED"];
+    this.s3Signer = new S3Signer();
+    this.bucket = process.env["KURL_BUCKET"] || "kurl-dev";
+  }
+
+  public async runCreateAirgapBundleJob(i: Installer) {
```
Comment (areed) [inline web/src/kubernetes/index.ts]: First create a ConfigMap with three scripts:
1. the bash script to rebundle all the individual packages
2. the rendered install.sh script
3. the rendered join.sh script

File: web/src/client/index_test.ts [hunk 10]
```diff
@@ -268,6 +321,29 @@ spec:
...
 `);
+  });
+
+  describe("/installer/latest?resolve=true", () => {
+    before(async () => {
+      await client.postInstaller(latest);
+    });
+
+    it("returns yaml with version", async () => {
+      const yaml = await client.getInstallerYAML("latest", true);
```
Comment (divolgin) [inline web/src/client/index_test.ts]: doesn't look like that second parameter is used in `getInstallerYAML`

File: web/src/kubernetes/index.ts [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+      },
+    };
+    const { body: cm } = await coreV1Client.createNamespacedConfigMap(this.ns, configMap);
+    const name = cm.metadata ? cm.metadata.name : genName;
+    const packages = i.packages().map((pkg) => `${this.distOrigin}/dist/${pkg}.tar.gz`);
+
+
+    const { signedUrl } = await this.s3Signer.signPutRequest({
+      Bucket: this.bucket,
+      Key: `bundle/${i.id}.tar.gz`,
```
Comment (divolgin) [inline web/src/kubernetes/index.ts]: is it worth checking here if the file already exists in s3 and then just not running the job if it does?

File: web/src/util/persistence/s3.ts [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+  }
+
+  public baseURL(params: GetParams): string {
+    return `https://${params.Bucket}.s3.amazonaws.com/${params.Key}`;
+  }
+
+  public parse(url: string): GetParams {
+    const parsed = new URL(url);
+    const Bucket = parsed.host.split(".")[0];
+    const Key = parsed.pathname.replace(/\//, "");
```
Comment (divolgin) [inline web/src/util/persistence/s3.ts]: not sure if this function is used, and what the purpose of this is, but slashes are an important part of the key in S3.

File: web/src/kubernetes/index.ts (hunk 11 above)
Comment (areed) [reply web/src/kubernetes/index.ts]: For named installers belonging to teams the bundle will need to be replaced.

File: web/src/util/persistence/s3.ts (hunk 12 above)
Comment (areed) [reply web/src/util/persistence/s3.ts]: not used, I'll delete

File: web/src/client/index_test.ts (hunk 10 above)
Comment (areed) [reply web/src/client/index_test.ts]: I forgot to implement that in the client. Added now.

**Pull request merged**